style-guide.md

```diff
@@ -0,0 +1,51 @@
+  - These texts should be copy and pasted from other tasks with the same input when possible
+    - see `template/common-parameter-meta.txt` for common description strings.
+  - Inputs and parameter meta must be in the same order
+  - That order should be logical
+    - No hard and fast rules, but look to similar tasks/workflows for an example of proper ordering
+    - Generally speaking:
+      - Required inputs at the top
+      - "sample" files before reference files
+      - Resource configuration at the bottom
+      - `max_retries` should be last
+      - If applicable, `detect_nproc` immediately before `max_retries`
+      - If applicable, use the same parameter name, help string, and parameter ordering as the underlying tool called by the task
+- Tasks with string parameters for which a limited number of choices are valid, must be documented following the template in `string_choices_task` (see `template/task-templates.wdl`)
+  - they should also fail quickly with an informative error message if an invalid input is provided
+    - In most cases, just passing the parameter to the underlying tool should produce a satisfactory error, but this must be checked for each task
+- All tasks must have configurable memory and disk space allocations
+  - see the various tasks in the template directory for possible ways to allocate resources
+    - Contributors can mix and match the available templates, copy and pasting subsections as appropriate
+    - It is allowed to have one resource allocated dynamically, and another allocated statically in the same task.
+    - It is not allowed to have a resource which can be allocated *either* statically or dynamically.
+      - This is technically feasible, but is too complicated for maintenance and end-users.
+- All tasks should have a `max_retries` input.
+  - This should be defaulted to `1` for nearly all tasks
+  - Some tasks are particularly error prone and can have a higher default `max_retries`
+  - **rule specific to workflows:** `max_retries` should be an optional `Int?`
+    - This allows each task to have it's own specific default `max_retries`
+      - If a user does not supply `max_retries`, those task level defaults will get used
+      - If a user does supply `max_retries`, it should override the default for *every* task called
+- multi-core tasks should always follow the conventions laid out in the `detect_nproc_task` example (see `template/task-templates.wdl`)
+- `command` blocks should be wrapped with arrows (`<<< >>>`) instead of brackets (`{ }`)
+  - Certain Bash constructions cause problems with the bracket notation
+- output file names should *always* be determined with either the `outfile_name` parameter or the `prefix` parameter.
+  - TODO: address tasks with multiple outputs
+- output variable names should be short but descriptive
+- All tasks should run in a Docker container
+  - TODO: provide a default Docker image to use
```

Comment:
Maybe we don't need a default. Given some more thought, I think each task has such specific needs that it becomes a futile effort to suggest a default. I think maybe instead of suggesting a default, I add the following bullet points:

```
- whenever possible, prefer an image maintained by an external source (such as BioContainers) to creating your own
- general purpose tasks can use the `util` image maintained in this repo
```

Reply:
tangentially related: should we address how we version Docker images in this PR? Or is that out of scope?

I can't remember how we landed on the current versioning scheme, and I'm not a huge fan of it. But I'm also not sure just using the underlying tools version is appropriate, because we make changes/additions and those should be tracked in the version number as well. It's not uncommon to update our docker image without touching the underlying tool. Would we just overwrite the previous version? I don't like that.

So I'm not sure what the right approach is, I see pros and cons to the two obvious choices. I don't know which one wins my vote, or if there's another possibility I haven't thought of.

Reply:
Potential third option: It would require breaking from SemVer so I'm a little hesitant right off the bat. (Although some of the tools we use don't follow SemVer; so adopting the "version based on underlying tool" approach also means leaving SemVer).

I'm proposing adding another tracking number to the underlying tool, which represents "our" patch number for the tool.
For example, we use STAR version `2.7.10b` (not SemVer). Our Docker image would get tagged as `star:2.7.10b.0`. Then, say we make a change like updating `sort_star_input.py`, we'd bump the Docker tag to `star:2.7.10b.1`.
When we update the tools version, "our patch" number gets reset to `0`.

I think this is a good compromise that should leave everyone happy. It makes it explicit which version of the tool is being used, and also allows us to track changes for our own purposes.

What do you think @adthrasher ?

Reply:
As I was rereading this thread, your latest proposal was exactly what popped into my head. I think we would need to consider exactly how to tag the images, but incorporating the underlying tool version along with our own incrementable version seems like a good compromise. Maybe a Slack discussion with the genomics team would be warranted. I'm curious to hear what others, who have more experience versioning, think.

Reply:
Sounds good to me. I'll writeup a summary and post it in the genomics channel!